src/libcore/str/mod.rs

```diff
@@ -474,58 +408,146 @@ impl<'a> FnOnce<(&'a u8,)> for BytesDeref {
     }
 }
 
-/// An iterator over the substrings of a string, separated by `sep`.
-struct CharSplits<'a, P: Pattern<'a>> {
-    /// The slice remaining to be iterated
-    start: usize,
-    end: usize,
-    matcher: P::Searcher,
-    /// Whether an empty string at the end is allowed
-    allow_trailing_empty: bool,
-    finished: bool,
+#[stable(feature = "rust1", since = "1.0.0")]
+impl<'a> Iterator for Bytes<'a> {
+    iter_methods!(: u8);
 }
 
-/// An iterator over the substrings of a string, separated by `sep`,
-/// splitting at most `count` times.
-struct CharSplitsN<'a, P: Pattern<'a>> {
-    iter: CharSplits<'a, P>,
-    /// The number of items remaining
-    count: usize,
+#[stable(feature = "rust1", since = "1.0.0")]
+impl<'a> DoubleEndedIterator for Bytes<'a> {
+    iter_methods!(double: u8);
 }
 
-/// An iterator over the substrings of a string, separated by a
-/// pattern, in reverse order.
-struct RCharSplits<'a, P: Pattern<'a>> {
-    /// The slice remaining to be iterated
-    start: usize,
-    end: usize,
-    matcher: P::Searcher,
-    /// Whether an empty string at the end of iteration is allowed
-    allow_final_empty: bool,
-    finished: bool,
+#[stable(feature = "rust1", since = "1.0.0")]
+impl<'a> ExactSizeIterator for Bytes<'a> {
+    iter_methods!(exact);
 }
 
-/// An iterator over the substrings of a string, separated by a
-/// pattern, splitting at most `count` times, in reverse order.
-struct RCharSplitsN<'a, P: Pattern<'a>> {
-    iter: RCharSplits<'a, P>,
-    /// The number of splits remaining
-    count: usize,
+/// This macro generates a Clone impl for string pattern API
+/// wrapper types of the form X<'a, P>
+macro_rules! derive_pattern_clone {
+    ($t:ident, $s:ident, $e:expr) => {
+        impl<'a, P: Pattern<'a>> Clone for $t<'a, P>
+            where P::Searcher: Clone
+        {
+            fn clone(&self) -> Self {
+                let $s = self;
+                $e
+            }
+        }
+    }
 }
 
-/// An iterator over the lines of a string, separated by `\n`.
-#[stable(feature = "rust1", since = "1.0.0")]
-pub struct Lines<'a> {
-    inner: CharSplits<'a, char>,
+/// This macro generates iterator wrappers for string iterators that use the Pattern API.
+/// It is a macro because correctly dealing with forward/reverse and single/double ended
+/// involes a lot of redunant and error prone impls.
+///
+/// If P: Pattern, then 2x2 iterator impls need to be generated,
+/// 2 for forward/backward and 2 for single/double:
+///
+///                                  => forward wrapper: Iterator            with next_match
+/// P::Searcher: DoubleEndedSearcher => forward wrapper: DoubleEndedIterator with next_match_back
+/// P::Searcher: ReverseSearcher     => reverse wrapper: Iterator            with next_match_back
+/// P::Searcher: DoubleEndedSearcher => reverse wrapper: DoubleEndedIterator with next_match
+macro_rules! generate_pattern_iterator {
```

Comment:
I understand, but I think this macro is worth it - there where/are many bugs/maintenance problems over time caused by the different types and iterator implementations drifting apart in asymmetrical ways, or cases being missed, etc.

However, if there is anything I could add to make the behavior of these macros more obvious I'd be glad to add it.